[FSDP] Register state_dict hooks for FlatParamsWrapper even if params_list is empty

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74860

These pre/post hooks must be registered even if the FlatParamsWrapper does not flatten any parameters; the FlatParamsWrapper may contains some FSDP-wrapped submodules.

Differential Revision: [D35194483](https://our.internmc.facebook.com/intern/diff/D35194483/)

fegin committed:
[FSDP] Register state_dict hooks for FlatParamsWrapper even if params_list is empty

These pre/post hooks must be registered even if the FlatParamsWrapper does not flatten any parameters; the FlatParamsWrapper may contains some FSDP-wrapped submodules.

Differential Revision: [D35194483](https://our.internmc.facebook.com/intern/diff/D35194483/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Register state_dict hooks for FlatParamsWrapper even if params_list is empty"


These pre/post hooks must be registered even if the FlatParamsWrapper does not flatten any parameters; the FlatParamsWrapper may contains some FSDP-wrapped submodules.

Differential Revision: [D35194483](https://our.internmc.facebook.com/intern/diff/D35194483/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM but can we add the test described in https://github.com/pytorch/pytorch/issues/74810?

fegin committed:
Update on "[FSDP] Register state_dict hooks for FlatParamsWrapper even if params_list is empty"


These pre/post hooks must be registered even if the FlatParamsWrapper does not flatten any parameters; the FlatParamsWrapper may contains some FSDP-wrapped submodules.

Differential Revision: [D35194483](https://our.internmc.facebook.com/intern/diff/D35194483/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -274,6 +276,42 @@ def test_state_dict_load_into_local_module(self):
+                setattr(module, LINEAR_SKIP, linear_skip)
+                return fsdp, linear_skip_tensor_names
+
+        fsdp, linear_skip_tensor_names = _create_module()
```
`linear_skip_tensor_names` is unused. I think it is for checking unused parameter in checkpoint, can you add a TODO here?

rohan-varma commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -274,6 +276,42 @@ def test_state_dict_load_into_local_module(self):
+        loss = fsdp(inp)
+        loss.sum().backward()
+
+        state_dict = fsdp.state_dict()
```
I think the overall goal to resolve the issue is to ensure that this can be loaded into a local module, I can try this though and add a follow up change. Or feel free to do so if you have the time.

fegin committed:
Update on "[FSDP] Register state_dict hooks for FlatParamsWrapper even if params_list is empty"


These pre/post hooks must be registered even if the FlatParamsWrapper does not flatten any parameters; the FlatParamsWrapper may contains some FSDP-wrapped submodules.

Differential Revision: [D35194483](https://our.internmc.facebook.com/intern/diff/D35194483/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Register state_dict hooks for FlatParamsWrapper even if params_list is empty"


These pre/post hooks must be registered even if the FlatParamsWrapper does not flatten any parameters; the FlatParamsWrapper may contains some FSDP-wrapped submodules.

Differential Revision: [D35194483](https://our.internmc.facebook.com/intern/diff/D35194483/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Register state_dict hooks for FlatParamsWrapper even if params_list is empty"


These pre/post hooks must be registered even if the FlatParamsWrapper does not flatten any parameters; the FlatParamsWrapper may contains some FSDP-wrapped submodules.

Differential Revision: [D35194483](https://our.internmc.facebook.com/intern/diff/D35194483/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Register state_dict hooks for FlatParamsWrapper even if params_list is empty"


These pre/post hooks must be registered even if the FlatParamsWrapper does not flatten any parameters; the FlatParamsWrapper may contains some FSDP-wrapped submodules.

Differential Revision: [D35194483](https://our.internmc.facebook.com/intern/diff/D35194483/)

[ghstack-poisoned]
